sdk/core/azure-core/src/main/java/com/azure/core/util/polling/Poller.java

```diff
@@ -305,7 +305,9 @@ public R block() {
         if (!isAutoPollingEnabled()) {
             setAutoPollingEnabled(true);
         }
-        this.fluxHandle.blockLast();
```

Comment:
Lots of good observations here. Let's go through them one-by-one:

1. Sounds reasonable
2. Sounds reasonable
3. I think a) sounds best. The b) option might not be obvious to users, especially in the sync case.
4. Clean up!
5. Please fix.
6. Please fix.

Reply:
thanks @JonathanGiles for this review. Agree with your take on all the items!

Reply:
**Observation7**:

With our current implementation, The cancellation signal is not propagated to the polling logic and to active subscribers.

A call to `Poller::cancelOperation()` should be considered as request to:

1. Invocation of original cancel method (that user provided when creating Poller) and return cancellation handle (a.k.a `Cancel Mono<T>`) to the caller of `Poller::cancelOperation()`.
2. Stop the polling once `Cancel Mono<T>` emit a signal followed by the subscription.
3. Inform all subscribers about cancel event.

***Suggestion:***

1. We could signal the cancellation event to active subscribers and future subscribers as an instance of `PollResponse` e.g. `PollResponse(OperationStatus.USER_CANCELLED, null)` a.k.a POLL_CANCELLED_RESPONSE.
2. Since we use `OperationStatus.USER_CANCELLED`, it has `OperationStatus::completed` set to true.
3. Since `OperationStatus::completed == true`, poller stops polling and treat POLL_CANCELLED_RESPONSE as terminal response.
4. Future subscribers are automatically taken care since we do `cache(1)` i.e. we cache the latest response, in this case POLL_CANCELLED_RESPONSE and replay it.


```java
public class Poller<T, R> {
    private volatile boolean operationCancelled;
    private static final PollResponse<T> POLL_CANCELLED_RESPONSE
        = new PollResponse<>(OperationStatus.USER_CANCELLED, null);

    public Mono<T> cancelOperation() {
        // .... [existing logic]

        //Time to call cancel
        return this.cancelOperation.apply(this).doFinally(() -> {
            operationCancelled = true;
        });
    }

    private Mono<PollResponse<T>> asyncPollRequestWithDelay() {
        return Mono.defer(() -> { 
            if (operationCancelled) {
                return POLL_CANCELLED_RESPONSE;
            }
            return this.pollOperation.apply(this.pollResponse)
                .delaySubscription(getCurrentDelay())
                .onErrorResume(throwable -> {
                    // ..
                    return Mono.empty();
                })
                .doOnEach(pollResponseSignal -> {
                    if (pollResponseSignal.get() != null) {
                        this.pollResponse = pollResponseSignal.get();
                    }
                });
            });
    }

}
```

***Assumptions:***

1. It's the caller that receives `cancel Mono<T>` is responsible for performing any needed retry in case `cancel Mono<T>` emits error.
2. Whether the caller handles the error or supress the error, the poller treat a terminal signal from `cancel Mono<T>` as cancellation of polling.

@JonathanGiles please review above cancel behavior and assumptions.

Reply:
**Observation8**

This is nit; and nice to have for better predictability.

The poll block looks like:

```java
private Mono<PollResponse<T>> asyncPollRequestWithDelay() {
    return Mono.defer(() -> { 
        return this.pollOperation.apply(this.pollResponse)
            .delaySubscription(getCurrentDelay())
            .onErrorResume(throwable -> {
                // ..
                return Mono.empty();
            })
            .doOnEach(pollResponseSignal -> {
                if (pollResponseSignal.get() != null) {
                    this.pollResponse = pollResponseSignal.get();
                }
            });
        });
}
```

the repeat() operator that call  `asyncPollRequestWithDelay` will "immediately" subscribe to this `defer`, that cause immediate call into pollOperation then it applies delay on publisher returned by pollOperation.

It would be nice to apply delay on invocation of pollOperation method itself.

***suggestion:**

```java
private Mono<PollResponse<T>> asyncPollRequestWithDelay() {
    return Mono.defer(() -> { 
        return this.pollOperation.apply(this.pollResponse)
            .onErrorResume(throwable -> {
                // ..
                return Mono.empty();
            })
            .doOnEach(pollResponseSignal -> {
                if (pollResponseSignal.get() != null) {
                    this.pollResponse = pollResponseSignal.get();
                }
            });
        }).delaySubscription(getCurrentDelay());
}
```

Reply:
Agree with 7 and 8.

Reply:
The first observation about delayActivation is spot on, in that it causes the activationOperation to be invoked twice.. and failing a storage test. I hope you don't mind if I swap this order in my PR to fix storage tests.

Reply:
@conniey sure. 

\\\cc @g2vinay FYI.